### PR TITLE
nixos/ssh: use upstream's default cryptographic algorithms

### PR DESCRIPTION
**Pull request by tomfitzhenry**
## Motivation

Hardening SSH algorithms, which typically means dropping all-but-the-strongest is of questionable value, given SSH's downgrade protection[0]. We pay in compatibility, and maintenance.

Further, as noted in
https://github.com/NixOS/nixpkgs/pull/172393/files#r871727289 , both the guidelines that we follow have not been updated in years.

The costs of having/maintaining these defaults:

1. The burden of having a larger module that deviates from upstream. We've slowly been reducing the upstream diff, to reduce maintenance burden.
2. Difficult for users to opt-out of these defaults. For example, when using a "no OpenSSL" build of OpenSSH, having these defaults means having to manually overriding NixOS's defaults. Upstream's defaults, meanwhile, gracefully only use available algorithms, if OpenSSL is not linked.
    * For users seeking to reduce attack surfaces that are fortunate enough to only have modern clients, they could choose to use `pkgs.opensshPackages.openssh.override { linkOpenssl = false; }`, which only supports chacha20-poly1305 and curve25519-sha256. This has the added benefit of removing openssl from the memory of sshd.
3. https://github.com/NixOS/nixpkgs/pull/231165 unexpectedly broke many clients.
4. The time in discussing/reviewing these defaults.
5. Anecdotally, a friend trying NixOS for the first time with a ssh_config supporting only ecdh-* key exchanges was unable to SSH in after install.

There's a certain level of enjoyment that comes from researching and selecting a favourite suite of ciphers, but as a distro, it's not our core competancy, and best left for upstream who are active in advances/attacks/compatibility.

0. https://eprint.iacr.org/2016/072.pdf

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by aszlig** [comment]
I agree that we shouldn't deviate too much from upstream, but we could alternatively move those defaults to something like [`profiles/hardened.nix`](https://github.com/NixOS/nixpkgs/blob/c5b6c986b3c071de25a192be7989e1fbb9e9fef9/nixos/modules/profiles/hardened.nix) or even a dedicated module for people to opt-in.

**Review by SuperSandro2000 — COMMENTED** [review]
We should document in the options that null means upstreams defaults. Also we should add a release note entry. 

Did you do a comparison which algorithms are now enabled again? Is there maybe some upstream guidance we could use instead like those algorithms are soon to be disabled by default and if you can, you should already disable them?

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 1]
```diff
@@ -19,6 +19,8 @@
 - `openssh` and `openssh_hpn` are now compiled without Kerberos 5 / GSSAPI support in an effort to reduce the attack surface of the components for the majority of users. Users needing this support can
   use the new `opensshWithKerberos` and `openssh_hpnWithKerberos` flavors (e.g. `programs.ssh.package = pkgs.openssh_gssapi`).
 
+- The `openssh` module now uses upstream's default cryptographic algorithms, as part of a broad effort to reduce the upstream diff in response to the 2024 xz incident.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2411.section.md:22]:
```suggestion
- The `openssh` module now uses upstream's default cryptographic algorithms in an effort to reduce required maintenance.
```
I think mentioning xz here will just trigger unnecessary questions and lead people into a wrong direction.

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 2]
```diff
@@ -417,36 +417,13 @@ in
...
-              description = ''
-                Allowed key exchange algorithms
-
-                Uses the lower bound recommended in both
-                <https://stribika.github.io/2015/01/04/secure-secure-shell.html>
-                and
-                <https://infosec.mozilla.org/guidelines/openssh#modern-openssh-67>
-              '';
+              default = null;
+              description = "Allowed key exchange algorithms. null will use upstream's defaults.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:421]:
```suggestion
              description = "Allowed key exchange algorithms. `null` uses upstream's default.";
```

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 3]
```diff
@@ -417,36 +417,13 @@ in
...
-              description = ''
-                Allowed MACs
-
-                Defaults to recommended settings from both
-                <https://stribika.github.io/2015/01/04/secure-secure-shell.html>
-                and
-                <https://infosec.mozilla.org/guidelines/openssh#modern-openssh-67>
-              '';
+              default = null;
+              description = "Allowed MACs. null will use upstream's defaults.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:426]:
```suggestion
              description = "Allowed MACs. `null` uses upstream's default.";
```

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 4]
```diff
@@ -457,22 +434,8 @@ in
...
-              description = ''
-                Allowed ciphers
-
-                Defaults to recommended settings from both
-                <https://stribika.github.io/2015/01/04/secure-secure-shell.html>
-                and
-                <https://infosec.mozilla.org/guidelines/openssh#modern-openssh-67>
-              '';
+              default = null;
+              description = "Allowed ciphers. null will use upstream's defaults.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:438]:
```suggestion
              description = "Allowed ciphers. `null` uses upstream's default.";
```

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 5]
```diff
@@ -417,36 +417,13 @@ in
             };
             KexAlgorithms = mkOption {
               type = types.nullOr (types.listOf types.str);
-              default = [
-                "sntrup761x25519-sha512@openssh.com"
-                "curve25519-sha256"
-                "curve25519-sha256@libssh.org"
-                "diffie-hellman-group-exchange-sha256"
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:424]: The new ones are

```
sntrup761x25519-sha512@openssh.com,
curve25519-sha256,curve25519-sha256@libssh.org,
ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,
diffie-hellman-group-exchange-sha256,
diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,
diffie-hellman-group14-sha256
```

https://github.com/openssh/openssh-portable/blob/V_9_7_P1/sshd_config.5#L1053C1-L1058C30

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 6]
```diff
@@ -417,36 +417,13 @@ in
...
-              '';
+              default = null;
+              description = "Allowed key exchange algorithms. null will use upstream's defaults.";
             };
             Macs = mkOption {
               type = types.nullOr (types.listOf types.str);
-              default = [
-                "hmac-sha2-512-etm@openssh.com"
-                "hmac-sha2-256-etm@openssh.com"
-                "umac-128-etm@openssh.com"
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:440]:
```
umac-64-etm@openssh.com,umac-128-etm@openssh.com,
hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,
hmac-sha1-etm@openssh.com,
umac-64@openssh.com,umac-128@openssh.com,
hmac-sha2-256,hmac-sha2-512,hmac-sha1
```

https://github.com/openssh/openssh-portable/blob/V_9_7_P1/sshd_config.5#L1206-L1210

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 7]
```diff
@@ -457,22 +434,8 @@ in
             };
             Ciphers = mkOption {
               type = types.nullOr (types.listOf types.str);
-              default = [
-                "chacha20-poly1305@openssh.com"
-                "aes256-gcm@openssh.com"
-                "aes128-gcm@openssh.com"
-                "aes256-ctr"
-                "aes192-ctr"
-                "aes128-ctr"
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:466]: https://github.com/openssh/openssh-portable/blob/V_9_7_P1/sshd_config.5#L578-L580

```
chacha20-poly1305@openssh.com,
aes128-ctr,aes192-ctr,aes256-ctr,
aes128-gcm@openssh.com,aes256-gcm@openssh.com
```

**Review by SuperSandro2000 — COMMENTED** [review]
I think the last point from the review thread is, to move the old defaults to the hardened profile.

**Review by SuperSandro2000 — COMMENTED** [review]
To test this out I just picked this PR into my fork and wanted to deploy it but it failed because I use openssh in the initrd 

I think we need to add some conditions to that config file

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/initrd-ssh.nix#L153-L155

**Comment by SuperSandro2000** [comment]
I fixed evaluation with https://github.com/SuperSandro2000/nixpkgs/commit/02c08fcc8355c02b639cfc81a5be779aa466f8db but I didn't test it yet e2e

**Comment by SuperSandro2000** [comment]
> * keeping the overridden defaults in `hardened.nix` does still offer improved ergonomics for those that want to opt-out of those overrides, but not for `hardened.nix` users (who have to do the un-ergonomic "Ciphers = lib.mkForce [ ... ]".

We could fix that by using lib.mkDefault

I don#t think we need an extra option for that

@aszlig opinions?

**Comment by aszlig** [comment]
> Alternatives are to have an option `enableMozillaRecommendations` (or whoever we're trusting, and I'd like us to select _one_ guide, if any).

That would be similar to eg. `services.nginx.recommendedTlsSettings`, which might probably be even preferable to the `hardened.nix` module, since if we want this in the `hardened` module it could still be included there via `services.openssh.enableMozillaRecommendations` (if the maintainers (ping @joachifm and @emilazy) of `hardened.nix` feel inclined to do).

> We could fix that by using lib.mkDefault

With the `enableMozillaRecommendations` option it would also address that issue since overriding is a single `mkForce` away.

**Comment by mweinelt** [comment]
Mozilla has fired their security team who was in charge of config recommendations back in 2020.

**Comment by LeSuisse** [comment]
Hello,

The given arguments seem fair to me however I'm a bit worried of re-introducing weak cryptography primitives by default. This seems like a regression and I would not be surprised if we end up getting issues from people getting suddenly a worse reports from <INSERT_YOUR_FAVORITE_COMMERCIAL_AUTOMATIC_PENTEST_TOOL> (or things like `ssh-audit` as shown in this PR).

I'm mostly worried about 2 things:
* re-introducing things like `hmac-sha1` or `umac-64@openssh.com` to the list of MACs
* (to a lesser extend) adding additional kex algos based on DH

both are arguably not something that is future proof.

We might have a middle ground solution that makes things easier for users wanting to harden their systems while not making it worse for users relying on the default values: `sshd` allows you to add or remove items from its default values using the `+` or `-` characters at the beginning of the list.
For example, right now at the top of your PR I can have `services.openssh.settings.Macs = ["-hmac-sha1" "hmac-sha1-etm@openssh.com" "umac-64@openssh.com"];` and `hmac-sha1` / `hmac-sha1-etm@openssh.com` / `umac-64@openssh.com` will be removed from the available MACs (I agree it feels a bit like a hack :smile: and maybe we should expose an nicer way to control that?).

This approach would allow us to exclude the crypto primitives with the lowest security margins and pick up the new ones proposed by `openssh` (I'm assuming they will be sane...) when we upgrade and also not cause issues with the `linkOpenssl = false;` use case.

**Pull request closed without merging**